pkg/composer/serviceparser/serviceparser.go

```diff
@@ -193,6 +215,60 @@ func getMemLimit(svc compose.ServiceConfig) (types.UnitBytes, error) {
 	return limit, nil
 }
 
+func getGPUs(svc compose.ServiceConfig) (reqs []string, _ error) {
+	// "gpu" and "nvidia" are also allowed capabilities (but not used as nvidia driver capabilities)
+	// https://github.com/moby/moby/blob/v20.10.7/daemon/nvidia_linux.go#L37
+	capset := map[string]struct{}{"gpu": {}, "nvidia": {}}
+	for _, c := range nvidia.AllCaps() {
+		capset[string(c)] = struct{}{}
+	}
+	if svc.Deploy != nil && svc.Deploy.Resources.Reservations != nil {
+		for _, dev := range svc.Deploy.Resources.Reservations.Devices {
+			if len(dev.Capabilities) == 0 {
+				// "capabilities" is required.
+				// https://github.com/compose-spec/compose-spec/blob/74b933db994109616580eab8f47bf2ba226e0faa/deploy.md#devices
+				return nil, fmt.Errorf("service %s: specifying \"capabilities\" is required for resource reservations", svc.Name)
+			}
+
+			var requiresGPU bool
+			for _, c := range dev.Capabilities {
+				if _, ok := capset[c]; ok {
+					requiresGPU = true
+				}
+			}
+			if !requiresGPU {
+				continue
+			}
+
+			var e []string
+			if len(dev.Capabilities) > 0 {
+				e = append(e, fmt.Sprintf("capabilities=%s", strings.Join(dev.Capabilities, ",")))
+			}
+			if dev.Driver != "" {
+				e = append(e, fmt.Sprintf("driver=%s", dev.Driver))
+			}
+			if len(dev.IDs) > 0 {
+				e = append(e, fmt.Sprintf("device=%s", strings.Join(dev.IDs, ",")))
+			}
+			if dev.Count != 0 {
+				e = append(e, fmt.Sprintf("count=%d", dev.Count))
+			}
+
```

Comment:
`count` and  `device_ids` are mutually exclusive. we should define one field at a time. is it configured somewhere ?Start a reviewAdd single commentCancel

Reply:
This is check at https://github.com/containerd/nerdctl/pull/251/files#diff-9d2133247dbc7d059776b8e61aaeb758cde413729d8ed7eb45ba1febbdb43ca6R139-R141